### PR TITLE
Adding CUP button and CUP mode.

### PR DESCRIPTION
**Pull request by williamlemus**
https://bugs.launchpad.net/mixxx/+bug/1535468

The CUP mode has been added. It works the following way:
- Press, pause at the cue point
- release, play

It is available as a Cue mode, or as a button that can be learned.

### CODE REVIEW

File: src/engine/cuecontrol.cpp [hunk 1]
```diff
@@ -754,12 +762,56 @@ void CueControl::cueDenon(double v) {
     }
 }
 
+void CueControl::cuePlay(double v) {
+    // This is how CUP button works:
+    // If playing, press to go to cue and stop.
+	// If stopped, press to set as cue point.
+	// On release, start playing from cue point.
```
Comment (sblaisot) [inline src/engine/cuecontrol.cpp]: indentation problem between the first two lines and the next two lines.

**Comment by sblaisot** [comment]
Thanks fo this PR.

Did you sign the mixxx contributor agreement ? https://docs.google.com/a/mixxx.org/spreadsheet/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ

Please also update the mixxxcontrols wiki page when this PR will be merge.

**Comment by daschuer** [comment]
Thank you for this Pull Request. 

Unfortunately this branch does not merge cleanly with our master branch. It is a good practice to keep your own master branch equal with the upstream master branch of mixxxdj/mixxx.

This can be done from the command line like that: 
```
# copy your changes into a feature branch 
git checkout master
git checkout -b CUP_feature

# adding mixxxdj/mixxx as upstream, if not already done:
git remote add upstream https://github.com/mixxxdj/mixxx.git

# set your master to upstream master 
git fetch upstream
git checkout master
git reset upstream/master --hard

# rebase your changes onto your new master
git checkout CUP_feature
git rebase master
# this will conflict. meld is a nice tool to resolve these conflicts side by side 
sudo apt-get install meld 

# resolve conflicts:
git mergetool
# add changed files 
git add <changed files list>
git rebase --continue  

# publish your changes
git push 

# now you have to issue a new pull request here on GitHub for your new branch.
```

File: src/controllers/controlpickermenu.cpp [hunk 2]
```diff
@@ -175,6 +175,8 @@ ControlPickerMenu::ControlPickerMenu(QWidget* pParent)
                    tr("Cue button (CDJ mode)"), cueMenu);
     addDeckControl("play_stutter", tr("Stutter Cue"),
                    tr("Stutter cue"), cueMenu);
+    addDeckControl("cue_play", tr("Cue (CUP Mode)"),
+                       tr("Cue button (CUP mode)"), cueMenu);
```
Comment (daschuer) [inline src/controllers/controlpickermenu.cpp]: IMHO this  the CUP button is a button of its own, not a cue button in a cue mode. 
How about:
```
addDeckControl("cue_play", tr("CUP (Cue Play)),
                          tr("Goto cue point and play after release"), cueMenu);
```

File: src/engine/cuecontrol.cpp (hunk 1 above)
Comment (daschuer) [inline src/engine/cuecontrol.cpp]: > // If stopped, press to set as cue point.
I am not sure if that is desired. Do you have a reference for this feature?

File: src/engine/cuecontrol.cpp [hunk 3]
```diff
@@ -754,12 +762,56 @@ void CueControl::cueDenon(double v) {
...
+			cueSet(v);
+			// Just in case.
+			m_bPreviewing = false;
+
+			// If quantize is enabled, jump to the cue point since it's not
+			// necessarily where we currently are
+			if (m_pQuantizeEnabled->get() > 0.0) {
+				lock.unlock();  // prevent deadlock.
+				// Enginebuffer will quantize more exactly than we can.
+				seekAbs(m_pCuePoint->get());
```
Comment (daschuer) [inline src/engine/cuecontrol.cpp]: Why was this added? IMHO we have to consider quantization only id we are about to play.

File: src/engine/cuecontrol.cpp [hunk 4]
```diff
@@ -754,12 +762,56 @@ void CueControl::cueDenon(double v) {
...
+}
+
 void CueControl::cueDefault(double v) {
     double cueMode = m_pCueMode->get();
     // Decide which cue implementation to call based on the user preference
     if (cueMode == CUE_MODE_DENON || cueMode == CUE_MODE_NUMARK) {
         cueDenon(v);
-    } else {
+    }else if(cueMode == CUE_MODE_CUP){
+        cuePlay(v);
```
Comment (daschuer) [inline src/engine/cuecontrol.cpp]: Will you use this mode?  I am just asking, because I have found more occurrences where CUP replaces Play. 
http://forum.djtechtools.com/showthread.php?t=17634&p=156018&viewfull=1#post156018

If it replaces the CUE button, than the cue set feature from above makes sense, otherwise not. 
mmmm that weird, do we actually need two CUP buttons, On with cue set function and on without?

**Comment by daschuer** [comment]
http://www.reloop.com/reloop-digital-jockey-2-master-edition#
in Tutorial Video 9, they say that CUP also sets the cue point when paused.

**Comment by daschuer** [comment]
Not sure, but here it is nothing about setting a cue point with CUP
http://www.reloop.com/media/custom/upload/Reloop-Reloop_Beatmix2_VirtualDJ8_Operation_Guide.pdf

**Comment by daschuer** [comment]
OK, after searching the web some more, I think your implementation with cue set feature is just right.

**Comment by williamlemus** [comment]
I will make a new pull request(rebased from the upstream master branch) with the following changes:

- Change tab to spaces

- Reword the addDeckControl to: 
```

addDeckControl("cue_play", tr("CUP (Cue Play)),
                          tr("Goto cue point and play after release"), cueMenu);
```

-  Remove the quantization:
```
			// If quantize is enabled, jump to the cue point since it's not
			// necessarily where we currently are
			if (m_pQuantizeEnabled->get() > 0.0) {
				lock.unlock();  // prevent deadlock.
				// Enginebuffer will quantize more exactly than we can.
				seekAbs(m_pCuePoint->get());
```
